### PR TITLE
Adding AS4 connector

### PR DESCRIPTION
**Pull request by manoramahp**
This implementation is compliant with AS4 Minimal Client Conformance Profile.
Git issue : https://github.com/wso2/product-ei/issues/852

### CODE REVIEW

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/AS4ErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+public class AS4ErrorHandler {
+
+    private static final Log log = LogFactory.getLog(AS4ErrorHandler.class);
+
+    /**
+     * Generate AS4 error messages and set to {@link MessageContext}
+     * @param messageContext {@link MessageContext} object
+     * @param e {@link AS4Exception} object
+     */
+    public static void generateErrorMessage(MessageContext messageContext, AS4Exception e) {
```
Comment (milindaperera) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/AS4ErrorHandler.java]: let's rename AS4Exception e with proper name.

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/file/AS4FileWriter.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            for (PartInfo partInfo : partInfoList) {
+                // Write attachment
+                if(partInfo.getHref().startsWith(AS4Constants.ATTACHMENT_HREF_PREFIX)) {
+                    String cid = partInfo.getHref().substring(AS4Constants.ATTACHMENT_HREF_PREFIX.length());
+                    DataHandler dataHandler = axis2MsgContext.getAttachment(cid);
+                    File partFile = new File(dataInFolder + File.separator + cid);
+                    try {
+                        dataHandler.writeTo(new FileOutputStream(partFile));
+                    } catch (IOException e) {
+                        log.error(e);
```
Comment (milindaperera) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/file/AS4FileWriter.java]: Shall we log the error message describing whats failing here
As done below when creating AS4Exception (same message will work here I think)

Comment (milindaperera) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/file/AS4FileWriter.java]: same in line 73

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    /**
+     * Determine if the error need to be notified to the producer end
+     * @param pmode {@link PMode} object
+     * @return true if <ProcessErrorNotifyProducer>true</ProcessErrorNotifyProducer>
+     */
+    public static boolean reportErrorToProducer(PMode pmode) {
+
+        boolean notifyProducer = false;
+        if(pmode.getErrorHandling() != null) {
+            if(pmode.getErrorHandling().getReport() != null) {
```
Comment (milindaperera) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java]: we can merge these two if statements to one

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    /**
+     * Determine if the error need to be sent as response to the sending MSH
+     * @param pmode {@link PMode} object
+     * @return true if <AsResponse>true</AsResponse>
+     */
+    public static boolean isSendErrorAsResponse(PMode pmode) {
+
+        boolean notifyProducer = false;
+        if(pmode.getErrorHandling() != null) {
+            if(pmode.getErrorHandling().getReport() != null) {
```
Comment (milindaperera) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java]: we can merge these two if statements to one

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/AS4ErrorHandler.java (hunk 1 above)
Comment (manoramahp) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/AS4ErrorHandler.java]: Fixed in https://github.com/wso2/carbon-mediation/pull/842/commits/28dc8826d26664836b4c45b5207d5ddbfde0fb41

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/file/AS4FileWriter.java (hunk 2 above)
Comment (manoramahp) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/file/AS4FileWriter.java]: Fixed in https://github.com/wso2/carbon-mediation/pull/842/commits/888d7d1b5084b251f5697ea55d3fd77bdf38ebc5

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java (hunk 3 above)
Comment (manoramahp) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java]: Fixed in https://github.com/wso2/carbon-mediation/pull/842/commits/888d7d1b5084b251f5697ea55d3fd77bdf38ebc5

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java (hunk 4 above)
Comment (manoramahp) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java]: Fixed in https://github.com/wso2/carbon-mediation/pull/842/commits/888d7d1b5084b251f5697ea55d3fd77bdf38ebc5

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+import javax.xml.stream.XMLStreamException;
+import java.io.IOException;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+
+public class AS4Utils {
+
+    private static final Log log = LogFactory.getLog(AS4Utils.class);
+
+    public static OMNode getOMNode(final Marshaller marshaller, final Messaging messaging)
```
Comment (vinok88) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java]: Add a doc comment.

**Comment by vinok88** [comment]
Shall we include a Jira/Ticket id to the commit message

**Comment by manoramahp** [comment]
Created a new git issue for AS4 support
https://github.com/wso2/product-ei/issues/852

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java (hunk 5 above)
Comment (manoramahp) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java]: Added in https://github.com/wso2/carbon-mediation/pull/842/commits/943cc42afba8e08bec9dc5ef78b4df9664b688df

File: components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/test/java/org/wso2/carbon/mediation/connector/as4IntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import org.wso2.connector.integration.test.base.RestResponse;
+
+import java.rmi.RemoteException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * AS4 connector integration test
+ */
+public class as4IntegrationTest extends ConnectorIntegrationTestBase {
```
Comment (vinok88) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/test/java/org/wso2/carbon/mediation/connector/as4IntegrationTest.java]: Class name should start with a capital letter

Comment (manoramahp) [inline components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/test/java/org/wso2/carbon/mediation/connector/as4IntegrationTest.java]: Renamed the class

**Pull request merged**